app/components/file/browse/file-explorer/file-explorer-workspace.ts

```diff
@@ -0,0 +1,161 @@
+import { FileLoader, FileNavigator, FileTreeNode } from "app/services/file";
+import { log } from "app/utils";
+import { BehaviorSubject, Observable } from "rxjs";
+
+export interface FileSource {
+    name?: string;
+    navigator: FileNavigator;
+    openedFiles?: string[];
+}
+
+export interface OpenedFile {
+    path: string;
+    source: FileSource;
+    loader: FileLoader;
+}
+function sourcesFrom(data: FileNavigator | FileSource | FileSource[]): FileSource[] {
+    if (data instanceof FileNavigator) {
+        return [{
+            name: "Files",
+            navigator: data,
+        }];
+    } else if (Array.isArray(data)) {
+        return data;
+    } else {
+        return [data];
+    }
+}
+
+export interface CurrentNode {
+    source: FileSource;
+    path: string;
+    treeNode: FileTreeNode;
+}
+
+export class FileExplorerWorkspace {
+    public sources: FileSource[];
+    public openedFiles: Observable<OpenedFile[]>;
+    // public openedFolder: Observable<any>;
+    public currentSource: Observable<FileSource>;
+    public currentNode: Observable<CurrentNode>;
+
+    private _currentSource = new BehaviorSubject<FileSource>(null);
+    private _currentPath = new BehaviorSubject("");
+    private _openedFiles = new BehaviorSubject<OpenedFile[]>([]);
+    // private _openedFolder = new BehaviorSubject<any>(null);
+
+    constructor(data: FileNavigator | FileSource | FileSource[]) {
+        this.sources = sourcesFrom(data);
+        this.openedFiles = this._openedFiles.asObservable();
+        // this.openedFolder = this._openedFolder.asObservable();
```

Comment:
is this going to be used in the future? otherwise delete: 
public openedFolder: Observable<any>

Reply:
forgot to remove it